File: core/logger.py
import os
import sys
import logging
from datetime import datetime
from typing import Optional
from rich.console import Console
from rich.panel import Panel
from rich.table import Table
from rich.text import Text

if sys.platform == "win32":
    try:
        sys.stdout.reconfigure(encoding="utf-8", errors="replace")
        sys.stderr.reconfigure(encoding="utf-8", errors="replace")
    except Exception:
        pass

console = Console(force_terminal=True, legacy_windows=False)

AUDIT_LOG_PATH = "output/audit.log"


def init_audit_logger(log_path: str = AUDIT_LOG_PATH) -> logging.Logger:
    """Initialize persistent audit logger for ethical/compliance recording."""
    os.makedirs(os.path.dirname(log_path) or ".", exist_ok=True)
    
    logger = logging.getLogger("recon_audit")
    logger.setLevel(logging.INFO)
    
    # Check if handlers already exist to avoid duplicate logs
    if not logger.handlers:
        handler = logging.FileHandler(log_path, encoding="utf-8")
        formatter = logging.Formatter(
            "%(asctime)s [%(levelname)s] %(message)s", datefmt="%Y-%m-%d %H:%M:%S"
        )
        handler.setFormatter(formatter)
        logger.addHandler(handler)
    
    return logger


audit_logger = init_audit_logger()


def log_audit(action: str, target: str, details: str = "", status: str = "SUCCESS") -> None:
    """Record an audit trail event."""
    msg = f"ACTION={action} | TARGET={target} | STATUS={status} | DETAILS={details}"
    audit_logger.info(msg)


def print_banner(app_name: str = "SpecterRecon", version: str = "1.0.0") -> None:
    """Display visually appealing ASCII banner."""
    banner_text = f"""
[bold cyan]  ____                  _             ____                     [/bold cyan]
[bold cyan] / ___| _ __   ___  ___| |_ ___ _ __ |  _ \ ___  ___ ___  _ __ [/bold cyan]
[bold cyan] \___ \| '_ \ / _ \/ __| __/ _ \ '__|| |_) / _ \/ __/ _ \| '_ \ [/bold cyan]
[bold cyan]  ___) | |_) |  __/ (__| ||  __/ |   |  _ <  __/ (_| (_) | | | |[/bold cyan]
[bold cyan] |____/| .__/ \___|\___|\__\___|_|   |_| \_\___|\___\___/|_| |_|[/bold cyan]
[bold cyan]       |_|                                                     [/bold cyan]
[bold white]   -- Fast, Modular Network Recon & Vulnerability Scanner v{version} --[/bold white]
[dim]                 Yetkilendirilmis Guvenlik ve Lab Test Platformu[/dim]
"""
    console.print(Panel(banner_text.strip(), border_style="cyan", expand=False))


def log_info(msg: str) -> None:
    console.print(f"[bold blue][*][/bold blue] {msg}")


def log_success(msg: str) -> None:
    console.print(f"[bold green][+][/bold green] {msg}")


def log_warning(msg: str) -> None:
    console.print(f"[bold yellow][!][/bold yellow] {msg}")


def log_error(msg: str) -> None:
    console.print(f"[bold red][-][/bold red] {msg}")


def log_step(step_name: str) -> None:
    console.print(f"\n[bold magenta]========= [ {step_name.upper()} ] =========[/bold magenta]")


def print_hosts_table(hosts) -> None:
    """Print discovered hosts in a clean Rich table."""
    if not hosts:
        return
    table = Table(title="🌐 Keşfedilen Hostlar", border_style="cyan", header_style="bold cyan")
    table.add_column("IP Adresi", style="bold white")
    table.add_column("MAC Adresi", style="dim")
    table.add_column("Yöntem", style="cyan")
    table.add_column("Gecikme (Latency)", style="green")
    table.add_column("Durum", style="bold green")

    for h in hosts:
        lat = f"{h.latency_ms} ms" if h.latency_ms is not None else "-"
        table.add_row(
            h.ip,
            h.mac or "-",
            h.discovery_method,
            lat,
            h.state.upper()
        )
    console.print(table)


def print_ports_table(ports) -> None:
    """Print open ports in a clean Rich table."""
    if not ports:
        return
    table = Table(title="🔌 Açık Portlar", border_style="blue", header_style="bold blue")
    table.add_column("IP Adresi", style="white")
    table.add_column("Port/Protokol", style="bold green")
    table.add_column("Servis", style="bold yellow")
    table.add_column("Durum", style="green")
    table.add_column("Yanıt Süresi", style="dim")

    for p in ports:
        resp = f"{p.response_time_ms} ms" if p.response_time_ms is not None else "-"
        table.add_row(
            p.ip,
            f"{p.port}/{p.protocol.upper()}",
            p.service_name or "unknown",
            p.state.upper(),
            resp
        )
    console.print(table)


def print_services_table(services) -> None:
    """Print detected services & banners in a clean Rich table."""
    if not services:
        return
    table = Table(title="🏷️ Tanımlanan Servisler & Versiyonlar", border_style="green", header_style="bold green")
    table.add_column("Hedef", style="white")
    table.add_column("Servis", style="bold cyan")
    table.add_column("Versiyon / Başlık", style="bold yellow")
    table.add_column("Açıklama / Banner", style="dim")
    table.add_column("SSL", style="magenta")

    for s in services:
        ver = s.service_version or s.http_title or "-"
        banner = s.banner_raw[:45] + "..." if (s.banner_raw and len(s.banner_raw) > 45) else (s.banner_raw or s.service_description or "-")
        ssl_str = "🔒 Evet" if s.ssl_enabled else "Hayır"
        table.add_row(
            f"{s.ip}:{s.port}",
            s.service_name.upper(),
            ver,
            banner,
            ssl_str
        )
    console.print(table)


def print_vulns_table(vulns) -> None:
    """Print detected CVE vulnerabilities in a clean Rich table."""
    if not vulns:
        console.print("[bold green]✅ Taranan servislerde bilinen kritik CVE zafiyeti bulunamadı.[/bold green]")
        return
    table = Table(title="🛡️ Tespit Edilen Zafiyetler & CVE'ler", border_style="red", header_style="bold red")
    table.add_column("CVE ID", style="bold white")
    table.add_column("CVSS", style="bold")
    table.add_column("Şiddet", style="bold")
    table.add_column("Etkilenen Servis", style="cyan")
    table.add_column("Zafiyet Açıklaması", style="white")

    for v in vulns:
        # Color coding for severity
        if v.severity == "CRITICAL":
            sev_styled = "[bold white on red] CRITICAL [/]"
            cvss_styled = f"[bold red]{v.cvss_score}[/bold red]"
        elif v.severity == "HIGH":
            sev_styled = "[bold black on yellow] HIGH [/]"
            cvss_styled = f"[bold yellow]{v.cvss_score}[/bold yellow]"
        elif v.severity == "MEDIUM":
            sev_styled = "[bold black on cyan] MEDIUM [/]"
            cvss_styled = f"[cyan]{v.cvss_score}[/cyan]"
        else:
            sev_styled = f"[dim]{v.severity}[/dim]"
            cvss_styled = f"[dim]{v.cvss_score}[/dim]"

        desc = v.description[:70] + "..." if len(v.description) > 70 else v.description
        table.add_row(
            v.cve_id,
            cvss_styled,
            sev_styled,
            v.affected_service,
            desc
        )
    console.print(table)


def print_dir_findings_table(findings) -> None:
    """Print web fuzzing directory findings in a clean Rich table."""
    if not findings:
        return
    table = Table(title="📁 Web Dizin & Dosya Bulguları (Fuzzer)", border_style="yellow", header_style="bold yellow")
    table.add_column("Durum Kodu", style="bold")
    table.add_column("URL / Yol", style="bold white")
    table.add_column("Boyut", style="cyan")
    table.add_column("Başlık / Yönlendirme", style="dim")
    table.add_column("Kritiklik", style="bold")

    for f in findings:
        # Status code color
        if f.status_code == 200:
            status_str = f"[bold green]{f.status_code} OK[/bold green]"
        elif f.status_code in [301, 302, 307]:
            status_str = f"[bold yellow]{f.status_code} REDIR[/bold yellow]"
        elif f.status_code in [401, 403]:
            status_str = f"[bold cyan]{f.status_code} AUTH[/bold cyan]"
        else:
            status_str = f"{f.status_code}"

        info = f.title or (f"➜ {f.redirect_location}" if f.redirect_location else "-")
        tag = "[bold red]⚠️ HASSAS DOSYA[/bold red]" if f.is_sensitive else "[dim]Standart[/dim]"

        table.add_row(
            status_str,
            f.url,
            f"{f.content_length} B",
            info[:35],
            tag
        )
    console.print(table)

File: core/logger.go
package core

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/pterm/pterm"
)

const DefaultAuditLogPath = "output/audit.log"

// LogAudit writes an audit log line with timestamp and action to audit.log.
func LogAudit(action, target, details, status string) {
	if status == "" {
		status = "SUCCESS"
	}
	dir := filepath.Dir(DefaultAuditLogPath)
	_ = os.MkdirAll(dir, 0755)

	f, err := os.OpenFile(DefaultAuditLogPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()

	timestamp := time.Now().Format("2006-01-02 15:04:05")
	entry := fmt.Sprintf("%s [INFO] ACTION=%s | TARGET=%s | STATUS=%s | DETAILS=%s\n",
		timestamp, action, target, status, details)
	_, _ = f.WriteString(entry)
}

// PrintBanner displays the ASCII banner.
func PrintBanner(version string) {
	bannerText := `  ____                  _             ____                     
 / ___| _ __   ___  ___| |_ ___ _ __ |  _ \ ___  ___ ___  _ __ 
 \___ \| '_ \ / _ \/ __| __/ _ \ '__|| |_) / _ \/ __/ _ \| '_ \ 
  ___) | |_) |  __/ (__| ||  __/ |   |  _ <  __/ (_| (_) | | | |
 |____/| .__/ \___|\___|\__\___|_|   |_| \_\___|\___\___/|_| |_|
       |_|                                                     
   -- Universal Network & Application Security Assessment v` + version + ` --`

	pterm.Println()
	pterm.DefaultBox.WithTitle("SPECTERRECON").WithTitleTopCenter().
		WithBoxStyle(pterm.NewStyle(pterm.FgCyan)).
		Println(bannerText)
}

// LogStep prints step header.
func LogStep(stepName string) {
	pterm.Println()
	pterm.DefaultSection.WithStyle(pterm.NewStyle(pterm.FgCyan, pterm.Bold)).
		Println(fmt.Sprintf("▸ %s", strings.ToUpper(stepName)))
}

// LogInfo prints informational message.
func LogInfo(format string, a ...interface{}) {
	msg := fmt.Sprintf(format, a...)
	pterm.Info.Println(msg)
}

// LogSuccess prints success message.
func LogSuccess(format string, a ...interface{}) {
	msg := fmt.Sprintf(format, a...)
	pterm.Success.Println(msg)
}

// LogWarning prints warning message.
func LogWarning(format string, a ...interface{}) {
	msg := fmt.Sprintf(format, a...)
	pterm.Warning.Println(msg)
}

// LogError prints error message.
func LogError(format string, a ...interface{}) {
	msg := fmt.Sprintf(format, a...)
	pterm.Error.Println(msg)
}

// PrintDNSTable prints DNS enumeration and subdomain findings in a clean PTerm table.
func PrintDNSTable(findings []DNSFinding) {
	if len(findings) == 0 {
		return
	}
	tableData := pterm.TableData{
		{"Hostname / FQDN", "IP Adresi", "Kayıt Tipi", "Kaynak"},
	}
	for _, f := range findings {
		src := "A/AAAA Çözümleme"
		if f.Source == "subdomain_bruteforce" {
			src = "🔍 Subdomain Brute-Force"
		}
		tableData = append(tableData, []string{
			f.Hostname,
			f.IP,
			f.RecordType,
			src,
		})
	}
	pterm.Println()
	_ = pterm.DefaultTable.WithHasHeader().WithBoxed().WithHeaderStyle(pterm.NewStyle(pterm.FgLightMagenta, pterm.Bold)).
		WithData(tableData).Render()
}

// PrintHostsTable prints discovered hosts in a clean PTerm table.
func PrintHostsTable(hosts []HostInfo) {
	if len(hosts) == 0 {
		return
	}
	tableData := pterm.TableData{
		{"IP Adresi", "MAC Adresi", "Yöntem", "Gecikme (Latency)", "Durum"},
	}
	for _, h := range hosts {
		mac := h.MAC
		if mac == "" {
			mac = "-"
		}
		lat := "-"
		if h.LatencyMs != nil {
			lat = fmt.Sprintf("%.2f ms", *h.LatencyMs)
		}
		tableData = append(tableData, []string{
			h.IP,
			mac,
			h.DiscoveryMethod,
			lat,
			strings.ToUpper(h.State),
		})
	}
	pterm.Println()
	_ = pterm.DefaultTable.WithHasHeader().WithBoxed().WithHeaderStyle(pterm.NewStyle(pterm.FgLightCyan, pterm.Bold)).
		WithData(tableData).Render()
}

// PrintPortsTable prints open ports in a clean PTerm table.
func PrintPortsTable(ports []PortInfo) {
	if len(ports) == 0 {
		return
	}
	tableData := pterm.TableData{
		{"IP Adresi", "Port/Protokol", "Servis", "Durum", "Yanıt Süresi"},
	}
	for _, p := range ports {
		resp := "-"
		if p.ResponseTimeMs != nil {
			resp = fmt.Sprintf("%.2f ms", *p.ResponseTimeMs)
		}
		service := p.ServiceName
		if service == "" {
			service = "unknown"
		}
		tableData = append(tableData, []string{
			p.IP,
			fmt.Sprintf("%d/%s", p.Port, strings.ToUpper(p.Protocol)),
			service,
			strings.ToUpper(p.State),
			resp,
		})
	}
	pterm.Println()
	_ = pterm.DefaultTable.WithHasHeader().WithBoxed().WithHeaderStyle(pterm.NewStyle(pterm.FgLightBlue, pterm.Bold)).
		WithData(tableData).Render()
}

// PrintServicesTable prints detected services & banners in a clean PTerm table.
func PrintServicesTable(services []ServiceDetail) {
	if len(services) == 0 {
		return
	}
	tableData := pterm.TableData{
		{"Hedef", "Servis", "Versiyon / Başlık", "Açıklama / Banner", "SSL"},
	}
	for _, s := range services {
		ver := s.ServiceVersion
		if ver == "" {
			ver = s.HTTPTitle
		}
		if ver == "" {
			ver = "-"
		}

		banner := s.BannerRaw
		if len(banner) > 40 {
			banner = banner[:37] + "..."
		}
		if banner == "" {
			banner = s.ServiceDescription
		}
		if banner == "" {
			banner = "-"
		}

		sslStr := "Hayır"
		if s.SSLEnabled {
			sslStr = "🔒 Evet"
		}

		tableData = append(tableData, []string{
			fmt.Sprintf("%s:%d", s.IP, s.Port),
			strings.ToUpper(s.ServiceName),
			ver,
			banner,
			sslStr,
		})
	}
	pterm.Println()
	_ = pterm.DefaultTable.WithHasHeader().WithBoxed().WithHeaderStyle(pterm.NewStyle(pterm.FgGreen, pterm.Bold)).
		WithData(tableData).Render()
}

// PrintDirFindingsTable prints web fuzzing directory findings in a clean PTerm table.
func PrintDirFindingsTable(findings []DirFuzzFinding) {
	if len(findings) == 0 {
		return
	}
	tableData := pterm.TableData{
		{"Durum Kodu", "URL / Yol", "Boyut", "Başlık / Yönlendirme", "Kritiklik / Eşleşme"},
	}
	for _, f := range findings {
		info := f.Title
		if info == "" && f.RedirectLocation != "" {
			info = "➜ " + f.RedirectLocation
		}
		if info == "" {
			info = "-"
		}
		if len(info) > 35 {
			info = info[:32] + "..."
		}

		tag := "Standart"
		if f.IsSensitive {
			tag = "⚠️ HASSAS DOSYA"
		} else if f.WordlistMatched != "" {
			tag = fmt.Sprintf("📁 %s", f.WordlistMatched)
		}

		tableData = append(tableData, []string{
			strconv.Itoa(f.StatusCode),
			f.URL,
			fmt.Sprintf("%d B", f.ContentLength),
			info,
			tag,
		})
	}
	pterm.Println()
	_ = pterm.DefaultTable.WithHasHeader().WithBoxed().WithHeaderStyle(pterm.NewStyle(pterm.FgYellow, pterm.Bold)).
		WithData(tableData).Render()
}

// PrintSslTable displays SSL audit findings in a styled terminal table.
func PrintSslTable(findings []SslFinding) {
	if len(findings) == 0 {
		return
	}
	tableData := pterm.TableData{
		{"IP:Port", "Sertifika Konusu", "Son Kullanma", "Kalan Gün", "Risk/Notlar"},
	}
	for _, f := range findings {
		subj := f.Subject
		if subj == "" && len(f.SANs) > 0 {
			subj = f.SANs[0]
		}
		if subj == "" {
			subj = "(Alınamadı)"
		}

		days := strconv.Itoa(f.DaysUntilExpiry)
		if f.IsExpired {
			days = "SÜRESİ DOLMUŞ!"
		} else if f.ExpiryDate == "" {
			days = "-"
		}

		notes := strings.Join(f.Notes, "; ")
		if notes == "" {
			notes = "Güvenli / Geçerli Sertifika"
		}
		if len(notes) > 40 {
			notes = notes[:37] + "..."
		}

		expiry := f.ExpiryDate
		if expiry == "" {
			expiry = "-"
		}

		tableData = append(tableData, []string{
			fmt.Sprintf("%s:%d", f.IP, f.Port),
			subj,
			expiry,
			days,
			notes,
		})
	}
	pterm.Println()
	_ = pterm.DefaultTable.WithHasHeader().WithBoxed().WithHeaderStyle(pterm.NewStyle(pterm.FgLightCyan, pterm.Bold)).
		WithData(tableData).Render()
}

// PrintHttpAuditTable displays HTTP security audit findings in a terminal table.
func PrintHttpAuditTable(findings []HttpAuditFinding) {
	if len(findings) == 0 {
		return
	}
	tableData := pterm.TableData{
		{"URL", "Eksik Headerlar", "Tehlikeli Metodlar", "CORS/GrafQL", "Şiddet"},
	}
	for _, f := range findings {
		missing := strings.Join(f.MissingHeaders, ", ")
		if len(missing) > 30 {
			missing = missing[:27] + "..."
		}
		methods := strings.Join(f.DangerousMethods, ", ")
		corsGql := "-"
		if len(f.CORSIssues) > 0 {
			corsGql = "CORS Risk"
		}
		if f.GraphQLOpen {
			if corsGql != "-" {
				corsGql += " | GraphQL"
			} else {
				corsGql = "GraphQL Introspect"
			}
		}
		tableData = append(tableData, []string{
			f.URL,
			missing,
			methods,
			corsGql,
			f.Severity,
		})
	}
	pterm.Println()
	_ = pterm.DefaultTable.WithHasHeader().WithBoxed().WithHeaderStyle(pterm.NewStyle(pterm.FgLightBlue, pterm.Bold)).
		WithData(tableData).Render()
}

// PrintSummaryTable displays final executive scan summary.
func PrintSummaryTable(report CompleteScanReport) {
	pterm.Println()
	tableData := pterm.TableData{
		{"Metrik", "Değer"},
		{"Taranan Hedef", report.Target},
	}
	if report.ScanProfile != "" {
		tableData = append(tableData, []string{"Tarama Profili", report.ScanProfile})
	}
	if report.TotalDNSRecords > 0 {
		tableData = append(tableData, []string{"Çözümlenen DNS Kayıtları", strconv.Itoa(report.TotalDNSRecords)})
	}
	tableData = append(tableData, [][]string{
		{"Keşfedilen Hostlar", strconv.Itoa(report.TotalHosts)},
		{"Açık Portlar", strconv.Itoa(report.TotalOpenPorts)},
		{"Web Dizin Bulguları", strconv.Itoa(report.TotalFindings)},
		{"Toplam Süre", fmt.Sprintf("%.2f saniye", report.DurationSeconds)},
		{"Özet Dosyası (TXT)", "output/summary.txt"},
		{"HTML Rapor Dosyası", "output/report.html"},
	}...)

	_ = pterm.DefaultTable.WithHasHeader().WithBoxed().WithHeaderStyle(pterm.NewStyle(pterm.FgCyan, pterm.Bold)).
		WithData(tableData).Render()
}
